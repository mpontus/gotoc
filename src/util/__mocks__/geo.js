module.exports = jest.genMockFromModule('../geo');
module.exports.getRadiusForRegion = jest.fn();
