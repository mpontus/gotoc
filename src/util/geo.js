// @flow
import { computeDestinationPoint, getDistance } from 'geolib';

/**
 * Calculate latitude and longitude deltas for radius around the point.
 *
 * @param {number} latitude North to south coordinate
 * @param {number} longitude East to west coordinate
 * @param {number} radius Radius in meters
 * @returns {object} Deltas in form { latitude: number, longitude: number }
 */
export function getDeltasForRadius(
  latitude: number,
  longitude: number,
  radius: number,
) {
  const start = { latitude, longitude };

  // Calculate coordinates for points around the starting point
  const getRelativePoint = direction =>
    computeDestinationPoint(start, radius, direction);
  const [north, east, south, west] = [0, 90, 180, 270].map(getRelativePoint);

  // Calculate deltas in all directions
  const [deltaNorth, deltaSouth] = [north, south]
    .map(point => point.latitude - latitude)
    .map(Math.abs);
  const [deltaWest, deltaEast] = [east, west]
    .map(point => point.longitude - longitude)
    .map(Math.abs);

  // Choose maximum distance as the delta
  const latitudeDelta = Math.max(deltaNorth, deltaSouth);
  const longitudeDelta = Math.max(deltaEast, deltaWest);

  return {
    latitude: latitudeDelta,
    longitude: longitudeDelta,
  };
}

/**
 * Calcualte radius which encircles the region around specified point
 *
 * @param {number} latitude North to south coordinate of the starting point
 * @param {number} longitude East to west coordinate of the starting point
 * @param {number} latitudeDelta Latitude delta
 * @param {number} longitudeDelta Longitude delta
 * @returns {number} Radius in meters
 */
export function getRadiusForRegion(
  latitude: number,
  longitude: number,
  latitudeDelta: number,
  longitudeDelta: number,
) {
  const start = { latitude, longitude };
  const topLeft = {
    latitude: latitude - latitudeDelta / 2,
    longitude: longitude - longitudeDelta / 2,
  };
  const bottomRight = {
    latitude: latitude + latitudeDelta / 2,
    longitude: longitude + longitudeDelta / 2,
  };

  return Math.max(getDistance(start, topLeft), getDistance(start, bottomRight));
}
