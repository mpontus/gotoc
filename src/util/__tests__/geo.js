import { getDeltasForRadius, getRadiusForRegion } from '../geo';

// Errors to the south, north, and right above equator
const points = [
  ['Nuuk Center, Greenland', 64.177995, -51.739083],
  ['Maua, Kenya', 0.2355187, 37.934525],
  ['Wellington, New Zealand', -41.284526, 174.7712372],
];

describe('geo utils', () => {
  describe('getDeltasForRadius', () => {
    points.forEach(([name, latitude, longitude]) => {
      test(`covers area around ${name}`, () => {
        expect(getDeltasForRadius(latitude, longitude, 1000)).toMatchSnapshot();
      });

      test(`does not exceed area around ${name}`, () => {
        expect(getDeltasForRadius(latitude, longitude, 1000)).toMatchSnapshot();
      });
    });
  });

  describe('getRadiusForRegion', () => {
    points.forEach(([name, latitude, longitude]) => {
      test(`radius around ${name}`, () => {
        const latitudeDelta = 0.2;
        const longitudeDelta = 0.04;

        expect(
          getRadiusForRegion(
            latitude,
            longitude,
            latitudeDelta,
            longitudeDelta,
          ),
        ).toMatchSnapshot();
      });
    });
  });
});
