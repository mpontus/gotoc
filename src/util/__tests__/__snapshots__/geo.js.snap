// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geo utils getDeltasForRadius covers area around Maua, Kenya 1`] = `
Object {
  "latitude": 0.008983152841195219,
  "longitude": 0.008983228735225168,
}
`;

exports[`geo utils getDeltasForRadius covers area around Nuuk Center, Greenland 1`] = `
Object {
  "latitude": 0.008983152841196329,
  "longitude": 0.020623576901471097,
}
`;

exports[`geo utils getDeltasForRadius covers area around Wellington, New Zealand 1`] = `
Object {
  "latitude": 0.008983152841196329,
  "longitude": 0.011954546791031362,
}
`;

exports[`geo utils getDeltasForRadius does not exceed area around Maua, Kenya 1`] = `
Object {
  "latitude": 0.008983152841195219,
  "longitude": 0.008983228735225168,
}
`;

exports[`geo utils getDeltasForRadius does not exceed area around Nuuk Center, Greenland 1`] = `
Object {
  "latitude": 0.008983152841196329,
  "longitude": 0.020623576901471097,
}
`;

exports[`geo utils getDeltasForRadius does not exceed area around Wellington, New Zealand 1`] = `
Object {
  "latitude": 0.008983152841196329,
  "longitude": 0.011954546791031362,
}
`;

exports[`geo utils getRadiusForRegion radius around Maua, Kenya 1`] = `11279`;

exports[`geo utils getRadiusForRegion radius around Nuuk Center, Greenland 1`] = `11210`;

exports[`geo utils getRadiusForRegion radius around Wellington, New Zealand 1`] = `11240`;
